node/libs/concurrency/src/sync/prunable_queue.rs

```diff
@@ -0,0 +1,148 @@
+//! Prunable queue mimics a producer-consumer channel, in enabling
+//! waiting on the receiver side for new items, while providing pruning functionality.
+//! The pruning takes place whenever a new item is sent, based on a specified predicate.
+//!
+//! The queue facilitates the asynchronous return of processing result for each added item.
+//!
+//! The separation of `Sender` and `Receiver` is employed primarily because `Receiver` requires
+//! a mutable reference to the signaling channel, unlike `Sender`, hence making it undesirable to
+//! be used in conjunction.
+
+use crate::{
+    ctx::{self, Ctx},
+    oneshot,
+    sync::{self, watch},
+};
+use std::{collections::VecDeque, sync::Arc};
+use tokio::sync::Mutex;
+
+pub fn new<T, U>(
+    pruning_predicate: Box<dyn Sync + Send + Fn(&T, &T) -> bool>,
+) -> (Sender<T, U>, Receiver<T, U>) {
+    let queue: Mutex<VecDeque<(T, oneshot::Sender<U>)>> = Mutex::new(VecDeque::new());
+    // Internal signaling, to enable waiting on the receiver side for new items.
+    let (send, recv) = watch::channel(false);
+
+    let shared = Arc::new(Shared {
+        queue,
+        has_items_send: send,
+    });
+
+    let sender = Sender {
+        shared: shared.clone(),
+        pruning_predicate: pruning_predicate,
+    };
+
+    let receiver = Receiver {
+        shared: shared.clone(),
+        has_items_recv: recv,
+    };
+
+    return (sender, receiver);
+}
+
+pub struct Shared<T, U> {
+    queue: Mutex<VecDeque<(T, oneshot::Sender<U>)>>,
+    has_items_send: watch::Sender<bool>,
+}
+
+pub struct Sender<T, U> {
+    shared: Arc<Shared<T, U>>,
+    pruning_predicate: Box<dyn Sync + Send + Fn(&T, &T) -> bool>,
+}
+
+impl<T, U> Sender<T, U> {
+    pub async fn enqueue(&self, item: T) -> oneshot::Receiver<U> {
+        // Create oneshot channel for returning result asynchronously.
+        let (res_send, res_recv) = oneshot::channel();
+
+        let mut queue = self.shared.queue.lock().await;
+        queue.retain(|existing_item| !(self.pruning_predicate)(&existing_item.0, &item));
+        queue.push_back((item, res_send));
+
+        // Ignore sending error.
+        let _ = self.shared.has_items_send.send(true);
+
+        res_recv
+    }
+}
+
+pub struct Receiver<T, U> {
+    shared: Arc<Shared<T, U>>,
+    has_items_recv: watch::Receiver<bool>,
+}
+
+impl<T, U> Receiver<T, U> {
+    pub async fn dequeue(&mut self, ctx: &Ctx) -> ctx::OrCanceled<(T, oneshot::Sender<U>)> {
+        sync::wait_for(ctx, &mut self.has_items_recv, |has_items| *has_items).await?;
```

Comment:
nit: you can obtain the same effect if you wrapped the VecDeque into a watch, no?
Sth like `Arc<watch::Sender<VecDeque<T>>>`.

Reply:
It might be possible, but I don't think it's a good thing to have a `T` for `watch` that needs to be mutated.

Reply:
why not though?

Reply:
It doesn't seem straightforward, but more like a hack.

Reply:
It doesn't seem straightforward, but more like a hack. The standard usage is to send new values.

Reply:
Mutating instead of sending new values doesn't seem straightforward, but more like a hack.

Reply:
Mutating, instead of sending new values, doesn't seem straightforward, but more like a hack.

Reply:
watch is not a channel. In fact it just a name for a mutex with notifications.

Reply:
Maybe I'm missing something, but I didn't find any `&mut` access from the sender.

Reply:
https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html#method.send_if_modified

Reply:
Fixed, changes here: https://github.com/matter-labs/era-consensus/pull/56/commits/dd4f659e59ced72f76d0d190c667fb8bedd73a07